[로또] 철시 미션 제출합니다.

pkch93 opened the pull request:
잘부탁드립니다! :)

pkch93 committed:
docs: README.md

'로또 1단계 - 자동'에서 필요한 기능 목록 추가

pkch93 committed:
feat: LottoNumber.class

로또 번호를 구현한 LottoNumber 클래스 구현

pkch93 committed:
refactor: LottoNumber.class

1~45로 범위가 정해져 있으므로 성능상 이점을 위해 미리 1~45까지의 LottoNumber를
만든 후 매번 새로운 인스턴스를 만드는 것이 아닌 기존의 값을 가져오는 방식으로 변경

pkch93 committed:
refactor: LottoNumber.class

매직넘버 상수 변경 (1 -> LOTTO_START_NUMBER, 45 -> LOTTO_LAST_NUMBER)

pkch93 committed:
feat: Lotto.class

로또 기능 구현

pkch93 committed:
refactor: Lotto.class

Lotto 클래스의 numbers 타입을 List<Integer> -> List<LottoNumber>로 변경

pkch93 committed:
refactor: Lotto.class

매직넘버 상수화 (6 -> LOTTO_SIZE)

pkch93 committed:
feat: WinningNumber

당첨 번호를 관리하는 객체
사용자가 구매한 로또로 당첨 여부 기능

pkch93 committed:
refactor: LottoTest.class

lotto 객체의 hasNumber가 변함에 따라 테스트 리펙토링

pkch93 committed:
feat: Result.class

최종 결과 목록을 가지고 있는 Result 객체 구현
수익률 계산 기능 구현

pkch93 committed:
refactor: Result.class

매직넘버 상수화 (100.0 -> ONE_HUNDRED_PERCENT)

pkch93 committed:
feat: InputView.class, InputValidator.class

구입금액과 당첨 번호를 입력받는 기능 구현

pkch93 committed:
feat: BoughtLottos.class 구현

사용자가 구매 금액을 입력하면 1000 단위로 절삭하여 로또를 만들어내는 로직 구현
- InputView.class 입력 메세지 추가
- 로또를 생성 기능을 담당하는 LottoGenerator 구현

pkch93 committed:
refactor: Lotto.class, LottoNumber.class

- BoughtLottos 구현 과정에서 Integer가 아닌 LottoNumber 리스트를 받아 Lotto
생성하는 로직으로 변경
- Collections.shuffle 메서드를 활용하기 위해 LottoNumber 클래스의 Map 자료구조를
List로 변경
- 변경에 따른 테스트 리펙토링

pkch93 committed:
style: Result.class

생성자에 한줄 띄워쓰기 제거

pkch93 committed:
feat: ResultGenerator.class 구현

사용자가 산 로또와 당첨 번호를 비교하여 최종 결과를 만들어내는 ResultGenerator
구현

pkch93 committed:
refactor: WINNING_NUMBER_DELIMITER 접근제어자 변경

pkch93 committed:
feat: BoughtLottos.class

구매한 로또의 갯수를 알려주는 메서드 countOfBoughtLottos 구현

pkch93 committed:
refactor: Lotto.class

Lotto과 관리하고 있는 List<LottoNumber>를 반환하는 getNumbers() 구현

pkch93 committed:
refactor: Result.class

상수 MIN_PRIZE_NUMBER와 MAX_PRIZE_NUMBER의 접근제어자 private -> public으로
변경
getCountOfPrize를 수익률 계산 뿐 아니라 OutputView에서도 쓸 수 있을 것이라 판단하여
public으로 접근제어자 변경

pkch93 committed:
refactor: Prize

Prize의 일치 갯수에 접근하는 getCountOfNumber() 구현

pkch93 committed:
feat: OutputView 구현

pkch93 committed:
feat: ConsoleUILottoApplication 구현

pkch93 committed:
style: OutputView 포메팅 및 import 제거

pkch93 committed:
refactor: Result의 상수 MIN_PRIZE_NUMBER와 MAX_PRIZE_NUMBER를 Prize로 이동

- Result보다 Prize에 더 걸맞는 상수라고 판단하여 옮김
- Prize에 옮김에 따라 이를 사용하고 있는 WinningNumber, OutputView 수정

pkch93 committed:
fix: LottoNumber.class

45가 유효한 범위가 아니라는 bug 수정

pkch93 committed:
feat: InputView inputBonusBall 구현

보너스 볼 입력 기능 구현

pkch93 committed:
docs: lotto 2단계 미션 요구사항 및 구현 기능 목록 추가

pkch93 committed:
feat: BonusBall.class 구현

- 구현 과정에서 LottoNumber의 상수값 LOTTO_START_NUMBER와 LOTTO_LAST_NUMBER
접근 제어자 변경 (private -> default)

pkch93 committed:
refactor: LottoGenerator.class

Collections.sort를 List.sort로 변경

pkch93 committed:
feat: Prize 2등(보너스볼 일치) 추가

- Prize 변경에 따른 ResultTest 리펙토링

pkch93 committed:
feat: WinningNumber 보너스 볼 추가

- 보너스볼 추가에 따른 당첨 로직 리펙토링

pkch93 committed:
refactor: Result.class

수익률 계산 로직의 for문을 Prize.values()의 foreach로 리펙토링

pkch93 committed:
feat: OutputView.class

- 2등 결과까지 추가하여 출력 기능 구현

pkch93 committed:
feat: 보너스 볼 입력 기능 구현

pkch93 committed:
refactor: InputValidator의 하드코딩 값 상수화

보너스 볼 입력값 검증을 담당하는 inputValidateBonusBall의 하드코딩 값을 상수화
이와 관련하여 LottoNumber의 LOTTO_START_NUMBER와 LOTTO_LAST_NUMBER의 접근
제어자를 public으로 변경

pkch93 committed:
refactor: printPrizeData CamelCase 변경

pkch93 committed:
refactor: enum Prize 안쓰는 상수 제거

pkch93 committed:
docs: 로또 2단계 완로 후 코드 개선사항 추가

pkch93 committed:
refactor: BonusBall.class의 인스턴스 변수 타입을 int -> LottoNumber로 변환

pkch93 committed:
refactor: WinningNumber.class

WinningNumber의 인스턴스 winningNumbers를 WinningLotto로 변경
이를 위해 WinningLotto 구현, WinningLotto는 당첨 번호를 가지고 있으며
lotto의 매칭갯수를 알려주는 역할을 한다.

이과정에서 BonusBall이 로또의 번호에 있는지 여부는 Lotto 클래스의 역할이라고 판단
하여 hasBonusBall 메서드를 Lotto 클래스의 구현

pkch93 committed:
test: WinningLotto, Lotto

로직 변경하면서 테스트 추가

pkch93 committed:
refactor: OutputView.class의 printPrizeData()

pkch93 committed:
docs: 로또 3단계 요구사항 및 기능목록 추가

pkch93 committed:
feat: 3단계 요구사항 반영

사용자 입력에 따른 Exception 추가
수동 입력 기능 구현
수동 구매 로또에 대한 출력 반영

pkch93 committed:
style: 전체 파일 포메팅 및 사용하지 않는 모듈 import 제거

pkch93 committed:
refactor: 리펙토링에 따른 test 코드 수정

jihan805 commented on src/main/java/lotto/domain/generator/LottoGenerator.java:
```diff
@@ -0,0 +1,41 @@
+
+import static lotto.domain.LottoNumber.getAllLottoNumbers;
+
+public class LottoGenerator {
```
LottoGenerator 클래스 추출 👍 
로또 생성은 자동/수동 방식이 있어요.
둘다 로또를 생성한다는 공통점이 있는데 인터페이스로 충분히 추출할 수 있지 않을까요?

pkch93 replied:
안녕하세요 서머님! 좋은 피드백 감사합니다!!

궁금한 점이 있어서 이렇게 댓글 남깁니다!

위 LottoGenerator는 상태를 가질 필요가 없다고 판단하여 하나의 LottoGenerator를 통해
static 메서드로 자동 / 수동 방식을 구현했었습니다.
피드백으로 인터페이스를 말씀해주셨는데 이 경우는 각각 static 메서드가 아닌 일반 메서드를 override하여 구현하는 게 좋을 것 같다는 말씀이신가요??

jihan805 replied:
네!!

jihan805 commented on src/main/java/lotto/domain/BoughtLottos.java:
```diff
@@ -0,0 +1,53 @@
+    public static BoughtLottos buyLottos(final int price, List<Lotto> manualLottos) {
+        List<Lotto> lottos = new ArrayList<>(manualLottos);
+        int countOfBoughtManual = manualLottos.size();
+        for (int i = 0; i < price / 1000 - countOfBoughtManual; i++) {
```
상수 사용!

jihan805 commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,49 @@
+        this.numbers = numbers;
+    }
+
+    public int hasNumber(final LottoNumber number) {
```
has~, is~ 는 주로 boolean 리턴에 대해 사용해요!

jihan805 commented on src/main/java/lotto/domain/Prize.java:
```diff
@@ -0,0 +1,50 @@
+
+import lotto.domain.exception.NotMatchLottoPrizeException;
+
+public enum Prize {
```
로또 결과를 enum 이용해 구현 👍
보너스볼 여부를 판단하기 위해 bonus인지 아닌지를 판단하는 값을 추가해보면 어떨까요?

pkch93 replied:
서머님께서 bonus 값을 추가하라는 피드백을 주셔서 반영을 해봤는데 결국 똑같은 코드가 되어 이렇게 질문드립니다. 

```java
// 반영 이전 (보너스 판단 값 x)
if (SECOND.matchCount(countOfNumber) && hasBonus) {
    return SECOND;
}
// 반영 이후 (보너스 판단 값 requiredBonus 적용)
if (SECOND.requiredBonus == hasBonus && SECOND.matchCount(countOfNumber)) {
    return SECOND;
}
```
위와 같이 SECOND 하나에 대해서만 Bonus 값이 필요하기 때문에 동일한 로직으로 구현이 되는 것 같습니다.
제가 주신 피드백에 대한 활용방안을 잘 모르겠어서 그런데 어떤 활용방안이 있을 수 있을 지 여쭤봐도 될까요???

jihan805 replied:
enum의 bonus가 true인 경우에만, 보너스볼로 넘어온 hasBonus를 확인하도록 하면 특정 경우(SECOND)에 대한 조건문을 줄일 수 있을 것 같아요!

jihan805 commented on src/test/java/lotto/domain/BoughtLottosTest.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    @Test
+    void init() {
```
전체적인 테스트 작성 👍 
init보다는 어떤 테스트인지 나타내는 이름이 더 좋을 것 같아요!

jihan805 commented on src/test/java/lotto/domain/LottoNumberTest.java:
```diff
@@ -0,0 +1,24 @@
+    @Test
+    void lotto숫자가_45보다_큰_경우() {
+        assertThrows(InvalidLottoNumberException.class, () -> LottoNumber.getLottoNumber(46));
+    }
```
경계값을 이용해 테스트 작성 💯

jihan805 commented on src/main/java/lotto/ConsoleUILottoApplication.java:
```diff
@@ -0,0 +1,33 @@
+    private static final int ONE_THOUSAND = 1000;
+
+    public static void main(String[] args) {
+        int buyPrice = cuttingThousandPiece(inputBuyPrice());
```
Money와 같은 객체로 추출해보면 어떨까요?

jihan805 commented on src/main/java/lotto/ConsoleUILottoApplication.java:
```diff
@@ -0,0 +1,33 @@
+                inputLottoNumber(INPUT_WINNING_NUMBER_MESSAGE)), inputBonusBall());
+        Result result = ResultGenerator.generateResult(boughtLottos, winningNumber);
+        OutputView.printLottoResult(result, buyPrice);
+    }
```
전체적인 구현 👍 
가독성을 위해 묶을 수 있는 단위로 줄을 띄우면 어떨까요?

jihan805 commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,51 @@
+public class LottoNumber {
+    public static final int LOTTO_START_NUMBER = 0;
+    public static final int LOTTO_LAST_NUMBER = 45;
+    private static final List<LottoNumber> numbers = new ArrayList<>();
```
LOTTO_START_NUMBER를 1로 충분히 구현할 수 있을 것 같아요!

jihan805 reviewed (changes requested):
안녕하세요 철시님, 
전체적으로 잘 구현하셨어요 💯, 추가로 README에 구현 과정을 꼼꼼하게 기록한 부분이 굉장히 인상적이에요 👍 
몇가지 피드백 남겼으니 확인 부탁드릴게요 :)

pkch93 committed:
refactor: class BoughtLottos

- 하드코딩 값 상수화 (1000 -> BUY_PRICE)
- 가독성을 위해 자동으로 만들어야하는 로또의 갯수를 변수화

pkch93 committed:
refactor: class Lotto

Lotto의 hasNumber 메서드 명을 checkNumber로 변경

pkch93 committed:
refactor: class LottoNumber

LOTTO_START_NUMBER의 값을 0에서 1로 변경

pkch93 committed:
refactor: Lottosgenerator

LottoGenerator를 LottosGenerator로 변환
수동/자동으로 생성할 수 있도록 interface와 구현체 생성
변경에 따른 코드 변경

pkch93 committed:
refactor: enum Prize

Prize의 각 상수값에 보너스 일치 여부에 대한 값 추가

pkch93 committed:
refactor: enum Prize

requiredBonus를 적용했으나 이전 작업물과 로직이 같게 되어 Prize에 추상메서드로
Prize를 판단하도록 구현

pkch93 committed:
refactor: class Money

Money 클래스를 적용하여 buyPrice 관리

pkch93 committed:
style: 전체 코드 포메팅 및 안쓰는 import 모듈 제거

pkch93 commented:
주신 피드백 토대로 구현해보았습니다!
아직 여전히 Generator에 대해 의문을 가지고 있습니다,, 상태값을 가질 필요가 없을때 static 메서드로 작성해야하는지와 이 과정에서 공통점이 보인다면 interface로 공통기능을 추출하여 각각의 구현체를 만드는 게 나은지에 대해서요,,

enum에 대해서는 먼저 보너스 필요 여부를 값으로 넣어봤으나 이전과 로직이 같은 코드가 나와서 enum에 대해서 알아보고 고민해 본 결과 추상메서드를 사용하여 구현해보았습니다,,

이번에도 좋은 피드백 부탁드립니다 :) 감사합니다!

jihan805 commented on src/main/java/lotto/domain/BoughtLottos.java:
```diff
@@ -0,0 +1,55 @@
+import java.util.Objects;
+
+public class BoughtLottos {
+    public static final int BUY_PRICE = 1000;
```
상수 활용 👍 
금액은 단위를 표현하기 위해 1_000와 같이 언더바로 구분해요!
위와 같이 표현하면 금액이 늘어나더라도 쉽게 읽을 수 있으므로 변경하면 어떨까요?

jihan805 commented on src/main/java/lotto/domain/LottosGenerator.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.List;
+
+public interface LottosGenerator {
+    List<Lotto> generate();
```
로또 생성은 상태가 없기 때문에 static 메서드로 구현하셨다고 코멘트에 남겨주셨습니다! 💯 
하지만 기존의 LottosGenerator에 대해 조금 더 자세히 고민해보면 수동 생성 + 자동 생성 두가지를 모두 담당하고 있고 
이는 충분히 분리할 수 있지 않을까요!?
한단계 더 나아가면 수동/자동 생성은 모두 "생성"이라는 공통점이 있으므로 이를 인터페이스로 추출할 수 있고요!
각각 로또를 생성하기 위해 필요한 인자가 다릅니다(갯수, 입력받은 리스트) 
이 부분에 대해  생성자를 활용해 처리 한 부분 역시 👍 잘하셨습니다.  💯 

추가로 피드백 남겨보겠습니다!
자동/수동 로또를 통합해 하나의 완전한 로또를 생성하도록 BoughtLottos 클래스를 수정해보면 어떨까요? 
main로직을 더 깔끔하게 변경할 수 있을 것 같아요!!

jihan805 commented on src/main/java/lotto/view/validator/InputValidator.java:
```diff
@@ -0,0 +1,58 @@
+
+    public static void inputValidateManualBuyLottoCount(final int buyPrice, final int countOfManualBuyLotto) {
+        int maxBuyLotto = buyPrice / BoughtLottos.BUY_PRICE;
+        if (!(maxBuyLotto - countOfManualBuyLotto >= 0)) {
```
이 부분 역시 상수를 사용할 수 있을 것 같아요!

jihan805 commented on src/main/java/lotto/view/validator/InputValidator.java:
```diff
@@ -0,0 +1,58 @@
+    private static final Pattern WINNING_NUMBER_REGEX = Pattern.compile("^([0-9]{1,2}, )+[0-9]{1,2}$");
+
+    public static void inputValidateBuyPrice(final String buyPrice) {
+        if (!BUY_PRICE_REGEX.matcher(buyPrice).matches()) {
```
InputValidator를 통해 input 에 대한 validation 체크  👍 

꼭 고민해보아야 할 중요한 부분이라고 생각해요!
로또 한장의 금액이 1000원 이상이어야 한다 - Money 클래스 / 로또한장에 중복이 있으면 안된다 - Lotto 클래스
~ 등과 같이 각 클래스에서 담당해야하는 벨리데이션도 섞여 있다고 생각해요.
Validator를 이용하는 것과 각 클래스가 역할에 맞는 벨리데이션을 책임지도록 하는 것은 어떤 차이가 있을까요?

pkch93 replied:
https://jojoldu.tistory.com/157

위 게시글을 통해 주신 피드백에 대해 학습하였습니다. 이번 미션에서 InputValidator는 프론트 단의 값을 처리하는 부분, 클래스가 역할에 맞는 벨리데이션을 가지는 것은 서버단이라고 생각하면서 이해해보았습니다.

현재 lotto 미션에서는 뷰에서 받은 값이 바로 Lotto나 Money등으로 전달되기 때문에 값이 변질되지는 않지만 프론트(InputView)에서 입력되는 값은 변질될 우려가 있기 때문에 실제 입력값으로 로직을 반영하는 클래스(서버)에서 예외에 따른 처리를 해줘야한다고 생각합니다.

좋은 피드백 감사합니다 :)

jihan805 commented on src/main/java/lotto/domain/generator/LottosAutoGenerator.java:
```diff
@@ -0,0 +1,39 @@
+    private Lotto generateAutoLotto() {
+        List<LottoNumber> allLottoNumbers = getAllLottoNumbers();
+        Collections.shuffle(allLottoNumbers);
+        List<LottoNumber> numbers = allLottoNumbers.subList(0, 6);
```
0~6 역시 상수를 사용할 수 있는 의미있는 숫자라고 생각해요!

jihan805 reviewed (approved):
안녕하세요 철시님!
전체적으로 피드백 반영 💯 
아주 간단한 코멘트 남겨두었으니 다음 단계 진행시 반영 부탁드릴게요!
머지 진행할게요~ 연휴 즐겁게 보내세요 :)
